src/expr/src/vector_op/split_part.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::array::{BytesGuard, BytesWriter};
+use risingwave_common::error::{ErrorCode, Result, RwError};
+
+#[inline(always)]
+pub fn split_part(
+    string_expr: &str,
+    delimiter_expr: &str,
+    nth_expr: i32,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    let mut split = string_expr.split(delimiter_expr);
+    let nth_val = if string_expr.is_empty() {
+        // postgres: return empty string for empty input string
+        Default::default()
+    } else if delimiter_expr.is_empty() {
+        // postgres: handle empty field separator
+        //           if first or last field, return input string, else empty string
+        if nth_expr == 1 || nth_expr == -1 {
+            string_expr
+        } else {
+            Default::default()
+        }
+    } else {
+        match nth_expr.cmp(&0) {
+            std::cmp::Ordering::Equal => {
+                return Err(RwError::from(ErrorCode::InvalidParameterValue(
+                    "field position must not be zero".into(),
+                )));
+            }
+            std::cmp::Ordering::Greater => split.nth(nth_expr as usize - 1).unwrap_or_default(),
+            std::cmp::Ordering::Less => {
+                let split = split.collect::<Vec<_>>();
+                let nth_expr = (split.len() as i32 + nth_expr) as usize;
+                split
+                    .get(nth_expr)
+                    .map_or(Default::default(), <&str>::clone)
+            }
+        }
+    };
+
+    let mut writer = writer.begin();
+    writer.write_ref(nth_val).unwrap();
+    Ok(writer.finish().unwrap())
+}
+
+#[cfg(test)]
+mod tests {
+    use risingwave_common::array::{Array, ArrayBuilder, Utf8ArrayBuilder};
+    use risingwave_common::error::{ErrorCode, Result, RwError};
+
+    use super::split_part;
+
+    #[test]
+    fn test_split_part() {
+        let cases: Vec<(&str, &str, i32, Result<&str>)> = vec![
+            ("", "@", 1, Ok("")),
+            ("", "@", -1, Ok("")),
+            ("joeuser@mydatabase", "", 1, Ok("joeuser@mydatabase")),
```

Comment:
Can we add a case `("@@@", "@@", -1, Ok("@"))`? This makes sure `rsplit` is not used internally when `nth` is negative.

Reply:
[da0ba96](https://github.com/singularity-data/risingwave/pull/2822/commits/da0ba968393cbf49f74ec904f9c4602c8eb3d8db)